pennylane/clusters.py

```diff
@@ -0,0 +1,476 @@
+              and returns NumPy arrays. It does not connect to any
+              machine learning library automatically for backpropagation.
+
+        diff_method (str, None): the method of differentiation to use in the created QNodeCluster.
+
+            * ``"best"``: Best available method. Uses the device directly to compute
+              the gradient if supported, otherwise will use the analytic parameter-shift
+              rule where possible with finite-difference as a fallback.
+
+            * ``"parameter-shift"``: Use the analytic parameter-shift
+              rule where possible with finite-difference as a fallback.
+
+            * ``"finite-diff"``: Uses numerical finite-differences.
+
+            * ``None``: a non-differentiable QNodeCluster is returned.
+
+    Returns:
+        QNodeCluster: a cluster of QNodes executing the circuit template with
+        the specified measurements
+
+    **Example:**
+
+    Let's define a custom two-wire template:
+
+    .. code-block:: python
+
+        def my_template(params, wires, **kwargs):
+            for i in range(2):
+                qml.RX(params[i], wires=wires[i])
+
+            qml.CNOT(wires=wires)
+
+    We want to compute the expectation values over the following list
+    of two-qubit observables:
+
+    >>> obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliX(1)]
+
+    This can be easily done with the ``map`` function:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> qnodes = qml.map(my_template, obs_list, dev, measure="expval")
+    >>> qnodes([0.54, 0.12])
+    array([-0.06154835  0.99280864])
+    """
+    if not callable(template):
+        raise ValueError("Could not create QNodes. The template is not a callable function.")
+
+    qnodes = QNodeCluster()
+
+    if not isinstance(device, Sequence):
+        # broadcast the single device over all observables
+        device = [device] * len(observables)
+
+    if not isinstance(measure, (list, tuple)):
+        # broadcast the single measurement over all observables
+        measure = [measure] * len(observables)
+
+    for obs, m, dev in zip(observables, measure, device):
+        # Generate QNodes from all pairs of observables, measurements, and devices.
+        if not isinstance(obs, Observable):
+            raise ValueError("Could not create QNodes. Some or all observables are not valid.")
+
+        wires = list(range(dev.num_wires))
+
+        def circuit(
+            params, _obs=obs, _m=m, _wires=wires, **kwargs
+        ):  # pylint: disable=dangerous-default-value, function-redefined
+            template(params, wires=_wires, **kwargs)
+            return MEASURE_MAP[_m](_obs)
+
+        qnode = QNode(circuit, dev, interface=interface, diff_method=diff_method)
+        qnodes.append(qnode)
+
+    return qnodes
+
+
+def apply(func, qnode_cluster):
+    """Lazily apply a function to the constituent QNodes of a :class:`QNodeCluster`.
```

Comment:
I'm a bit confused by what "lazily" means here, even after the explanation below saying that no QNode evaluation has occurred. Is "lazy" a standard term? Otherwise I would suggest removing "lazily" from the first line of the docstring, but keeping it below (line 178)

Reply:
I have to admit, I'm not to fond of the way I've phrased it either. But yes, it's a common term! https://en.wikipedia.org/wiki/Lazy_evaluation

Reply:
Feel free to suggest any alternate wordings though!

Reply:
I would suggest removing lazily, i.e., write "Apply a function to the constituent..." 
But keep note that it's lazy in line 178